src/workspace/resolver.rs

```diff
@@ -0,0 +1,373 @@
+// Copyright 2018-2024 the Deno authors. MIT license.
+
+use std::collections::BTreeMap;
+use std::future::Future;
+
+use anyhow::Error as AnyError;
+use deno_semver::npm::NpmPackageReqReference;
+use deno_semver::package::PackageReqReference;
+use import_map::specifier::SpecifierError;
+use import_map::ImportMap;
+use import_map::ImportMapDiagnostic;
+use import_map::ImportMapError;
+use import_map::ImportMapWithDiagnostics;
+use thiserror::Error;
+use url::Url;
+
+use crate::package_json::PackageJsonDepValueParseError;
+use crate::package_json::PackageJsonDeps;
+use crate::package_json::PackageJsonRc;
+use crate::sync::new_rc;
+use crate::ConfigFile;
+
+use super::UrlRc;
+use super::Workspace;
+
+#[derive(Debug)]
+struct PkgJsonResolverFolderConfig {
+  deps: PackageJsonDeps,
+  pkg_json: PackageJsonRc,
+}
+
+#[derive(Debug, Error)]
+pub enum WorkspaceResolverCreateError {
+  #[error("Failed loading import map specified in '{referrer}'")]
+  ImportMapFetch {
+    referrer: Url,
+    #[source]
+    source: AnyError,
+  },
+  #[error(transparent)]
+  ImportMap(#[from] import_map::ImportMapError),
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct CreateResolverOptions {
+  /// Whether to resolve dependencies by reading the dependencies list
+  /// from a package.json
+  ///
+  /// This should only be set when a node_modules folder doesn't exist
+  /// as node resolution should be used otherwise.
+  pub pkg_json_dep_resolution: bool,
```

Comment:
I don't understand this comment :( Does that mean it should only be used with "global resolver"?

Reply:
It should resolve based on node_modules folders instead of package.json. Actually, I just realized I think this will break for some npm specifiers with byonm. I'll look into it tomorrow.

Reply:
Ok, so basically what I mean by this is that we shouldn't resolve bare specifiers from the package.json, but instead by only looking at `node_modules` folder only. For example, if you create a package.json in node with a "ts-morph" dependency and install it, then delete ts-morph from the package.json's dependencies list, it should still resolve "ts-morph" because it exists in the node_modules folder.

Reply:
Can you update the comment?